### PR TITLE
Replace ring with hmac for hmacs

### PR DESCRIPTION
**Pull request by tomaka**
It's a bit of a painful birth because of #138, but this PR replaces `ring` with `hmac` for computing HMACs.
Part of #133

### CODE REVIEW

**Comment by ordian** [comment]
@tomaka could you clarify if the #138 applies to the current or the proposed implementation? Seems like the latter.
> Building one of these types with the wrong slice length panics.

This seems to be different from https://briansmith.org/rustdoc/ring/hmac/struct.SigningKey.html#method.new 
> As specified in RFC 2104, if key_value is shorter than the digest algorithm's block length (as returned by digest::Algorithm::block_len, not the digest length returned by digest::Algorithm::output_len) then it will be padded with zeros. Similarly, if it is longer than the block length then it will be compressed using the digest algorithm.

**Comment by tomaka** [comment]
#138 applies to both. I decided to not touch the API here.

> This seems to be different from https://briansmith.org/rustdoc/ring/hmac/struct.SigningKey.html#method.new

Oh, good remark. I really don't like this implicit behaviour in `ring`.

**Comment by cheme** [comment]
@tomaka I did write some compat for webassembly on most of crypto crate a while back, for your code it looks like I did use Hmac<H256> : https://github.com/cheme/parity-common/blob/crypto-compat/parity-crypto/src/hmac_alt.rs there is a test to compare both impl (the branch did keep both impl).

**Comment by ordian** [comment]
I like @cheme's version better, because [it can panic with a more meaningful message](https://docs.rs/generic-array/0.12.0/src/generic_array/lib.rs.html#548) and uses static dispatch. As a downside it adds a dependency on `hmac::Mac` for public types though.

**Comment by ordian** [comment]
> @ordian I am not sure having the Hmac is an issue (I did find back my test in ethereum paritytech/parity-ethereum@master...cheme:wasm-tests and it does not look like the api did change except for an import (for some other things it did change a lot eg secp where I add to put some curve math in a trait).
Edit: @ordian I think I misunderstood you, it does add some additional deps indeed (could probably find a way to avoid that).

I meant the fact, that we're using `hmac` crate is an implementation detail that shouldn't be exposed in the public API. That way we can migrate to other crate later w/o introducing breaking changes.

File: parity-crypto/src/hmac/mod.rs [hunk 1]
```diff
@@ -14,77 +14,119 @@
...
 	}
 }
 
 /// HMAC signing key.
-pub struct SigKey<T>(hmac::SigningKey, PhantomData<T>);
+pub struct SigKey<T>(KeyInner, PhantomData<T>);
+
+enum KeyInner {
+	Sha256(GenericArray<u8, U64>),
+	Sha512(GenericArray<u8, U128>),
```
Comment (ordian) [inline parity-crypto/src/hmac/mod.rs]: @tomaka should this be `U32` and `U64`?

Comment (tomaka) [reply parity-crypto/src/hmac/mod.rs]: Indeed, probably.

**Pull request merged**